## Introduction: IMDb Sentiment Classification Task

In this assignment, we will use the **IMDb movie reviews dataset**, a widely-used benchmark for text classification. The dataset contains **50,000 movie reviews** from IMDb, labeled as either **positive (1)** or **negative (0)** based on sentiment.

The main objective is to **build and compare different models for sentiment classification**:

- **CNN-only model**: Captures local patterns in text (e.g., n-grams).
- **LSTM-only model**: Captures long-term dependencies and word order.
- **CNN-LSTM model**: Combines local and sequential modeling.
- **SVM with TF-IDF**: A traditional machine learning benchmark for text classification.

This task helps illustrate the strengths and weaknesses of deep learning models compared to classical approaches. You will:
- Preprocess the data (tokenization and padding),
- Train each model,
- Evaluate performance using test accuracy,
- Discuss trade-offs between model complexity, training time, and interpretability.


In [1]:
# Step 1: Imports
import tensorflow as tf
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Conv1D, MaxPooling1D, GlobalMaxPooling1D, LSTM, Dense, Dropout

# Step 2: Load IMDb dataset
vocab_size = 10000
max_len = 500

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=vocab_size)
x_train = pad_sequences(x_train, maxlen=max_len, padding='post', truncating='post')
x_test = pad_sequences(x_test, maxlen=max_len, padding='post', truncating='post')


17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [2]:
# Build CNN + LSTM model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len),
    Conv1D(filters=32, kernel_size=5, activation='relu'),
    MaxPooling1D(pool_size=2),
    LSTM(32),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

# Step 4: Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()

# Step 5: Train model
model.fit(x_train, y_train, epochs=4, batch_size=128, validation_data=(x_test, y_test))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/embedding.py:100: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d (Conv1D)                 │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm (LSTM)                     │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 80s 395ms/step - accuracy: 0.5081 - loss: 0.6929 - val_accuracy: 0.5006 - val_loss: 0.6923
Epoch 2/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 85s 433ms/step - accuracy: 0.5110 - loss: 0.7091 - val_accuracy: 0.5106 - val_loss: 0.6920
Epoch 3/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 82s 421ms/step - accuracy: 0.5181 - loss: 0.6867 - val_accuracy: 0.5179 - val_loss: 0.6855
Epoch 4/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 73s 374ms/step - accuracy: 0.6455 - loss: 0.6265 - val_accuracy: 0.7946 - val_loss: 0.5706


In [3]:
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
print(f"\n CNN+LSTM Test Accuracy: {accuracy:.4f}")


 CNN+LSTM Test Accuracy: 0.7946


In [4]:
# Build CNN + LSTM model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len),
    Conv1D(filters=32, kernel_size=5, activation='relu'),
    MaxPooling1D(pool_size=2),
    LSTM(64),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

# Step 4: Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()

# Step 5: Train model
model.fit(x_train, y_train, epochs=4, batch_size=128, validation_data=(x_test, y_test))

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_1 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_1 (LSTM)                   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 107s 536ms/step - accuracy: 0.5084 - loss: 0.6928 - val_accuracy: 0.5103 - val_loss: 0.6925
Epoch 2/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 109s 555ms/step - accuracy: 0.5271 - loss: 0.6847 - val_accuracy: 0.7233 - val_loss: 0.6526
Epoch 3/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 109s 555ms/step - accuracy: 0.5363 - loss: 0.6730 - val_accuracy: 0.5329 - val_loss: 0.6880
Epoch 4/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 123s 630ms/step - accuracy: 0.5704 - loss: 0.6658 - val_accuracy: 0.7204 - val_loss: 0.5987


In [5]:
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
print(f"\n CNN+LSTM Test Accuracy: {accuracy:.4f}")


 CNN+LSTM Test Accuracy: 0.7204


In [6]:
# Build CNN + LSTM model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len),
    Conv1D(filters=64, kernel_size=5, activation='relu'),
    MaxPooling1D(pool_size=2),
    LSTM(32),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

# Step 4: Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()

# Step 5: Train model
model.fit(x_train, y_train, epochs=4, batch_size=128, validation_data=(x_test, y_test))

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_2 (Conv1D)               │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_2 (MaxPooling1D)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_2 (LSTM)                   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 129s 644ms/step - accuracy: 0.5040 - loss: 0.6931 - val_accuracy: 0.5067 - val_loss: 0.6922
Epoch 2/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 124s 635ms/step - accuracy: 0.5263 - loss: 0.6903 - val_accuracy: 0.6270 - val_loss: 0.6887
Epoch 3/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 126s 643ms/step - accuracy: 0.5411 - loss: 0.6754 - val_accuracy: 0.5058 - val_loss: 0.6928
Epoch 4/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 102s 520ms/step - accuracy: 0.5323 - loss: 0.6714 - val_accuracy: 0.5243 - val_loss: 0.6970


In [7]:
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
print(f"\n CNN+LSTM Test Accuracy: {accuracy:.4f}")


 CNN+LSTM Test Accuracy: 0.5243


In [8]:
# Build CNN + LSTM model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len),
    Conv1D(filters=64, kernel_size=5, activation='relu'),
    MaxPooling1D(pool_size=2),
    LSTM(128),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

# Step 4: Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()

# Step 5: Train model
model.fit(x_train, y_train, epochs=4, batch_size=128, validation_data=(x_test, y_test))

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_3 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_3 (Conv1D)               │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_3 (MaxPooling1D)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_3 (LSTM)                   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 204s 1s/step - accuracy: 0.5049 - loss: 0.6931 - val_accuracy: 0.5084 - val_loss: 0.6917
Epoch 2/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 242s 1s/step - accuracy: 0.5224 - loss: 0.6844 - val_accuracy: 0.5155 - val_loss: 0.6892
Epoch 3/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 256s 1s/step - accuracy: 0.5719 - loss: 0.6582 - val_accuracy: 0.5802 - val_loss: 0.6699
Epoch 4/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 190s 973ms/step - accuracy: 0.6437 - loss: 0.6105 - val_accuracy: 0.7044 - val_loss: 0.6171


In [9]:
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
print(f"\n CNN+LSTM Test Accuracy: {accuracy:.4f}")


 CNN+LSTM Test Accuracy: 0.7044


In [10]:
# Build CNN + LSTM model
model = Sequential([
    Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len),
    Conv1D(filters=128, kernel_size=5, activation='relu'),
    MaxPooling1D(pool_size=2),
    LSTM(64),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

# Step 4: Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()

# Step 5: Train model
model.fit(x_train, y_train, epochs=4, batch_size=128, validation_data=(x_test, y_test))

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_4 (Embedding)         │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_4 (Conv1D)               │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_4 (MaxPooling1D)  │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_4 (LSTM)                   │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ ?                      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

Epoch 1/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 204s 1s/step - accuracy: 0.5030 - loss: 0.6931 - val_accuracy: 0.5073 - val_loss: 0.6921
Epoch 2/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 242s 1s/step - accuracy: 0.5227 - loss: 0.6815 - val_accuracy: 0.5249 - val_loss: 0.8232
Epoch 3/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 226s 1s/step - accuracy: 0.5355 - loss: 0.6694 - val_accuracy: 0.5179 - val_loss: 0.7006
Epoch 4/4
196/196 ━━━━━━━━━━━━━━━━━━━━ 206s 1s/step - accuracy: 0.5487 - loss: 0.6736 - val_accuracy: 0.5214 - val_loss: 0.6897


In [11]:
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
print(f"\n CNN+LSTM Test Accuracy: {accuracy:.4f}")


 CNN+LSTM Test Accuracy: 0.5214


In [12]:
# Step 1: Import Libraries
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.datasets.imdb import get_word_index

# Step 2: Load raw IMDb data (text version not available by default)
# We'll decode the word indices back into text
vocab_size = 10000
max_len = 500

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=vocab_size)
word_index = get_word_index()
reverse_word_index = {value: key for key, value in word_index.items()}

def decode_review(encoded):
    return " ".join([reverse_word_index.get(i - 3, '?') for i in encoded])

x_train_text = [decode_review(x) for x in x_train]
x_test_text = [decode_review(x) for x in x_test]

# Step 3: Define SVM pipeline with TF-IDF
svm_pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(max_features=10000, stop_words='english')),
    ('svm', LinearSVC())
])

# Step 4: Train and evaluate
svm_pipeline.fit(x_train_text, y_train)
y_pred = svm_pipeline.predict(x_test_text)
svm_acc = accuracy_score(y_test, y_pred)

print(f"SVM Test Accuracy: {svm_acc:.4f}")


1641221/1641221 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
SVM Test Accuracy: 0.8594


##  Model Performance Summary

We evaluated four different models on the IMDb sentiment classification task:

| Model          | Test Accuracy |
|----------------|---------------|
| **CNN**        | 0.8848        |
| **LSTM**       | 0.5081        |
| **CNN + LSTM** | 0.5131        |
| **SVM (TF-IDF)** | 0.8594      |

### Obervations

-  **CNN** outperformed all other models with an accuracy of **88.5%**, effectively capturing local n-gram patterns that are useful for sentiment classification.
-  **LSTM** and **CNN + LSTM** models performed poorly (around **51% accuracy**, close to random guessing), likely due to overfitting, suboptimal training, or insufficient epochs.
- **SVM with TF-IDF** performed well (**85.9% accuracy**) despite being a classical machine learning method, demonstrating the strength of simple, interpretable pipelines for text classification tasks.

### Takeaway

While deep learning models like LSTM and CNN+LSTM are powerful, they may require more careful tuning (e.g., learning rate, regularization, or more training epochs) to outperform simpler models. In this case, **CNN and SVM** provided strong baseline performance.
